### PR TITLE
#3: Add specification and design

### LINKED ISSUE #3: Add specification and design
## Acceptance Criteria

* User requirement specification exists
* Design document exists
* Requirement tracing into implementation and tests is green

### PR DESCRIPTION
**Pull request by redcatbear**
Closes #3.

### CODE REVIEW

File: doc/design.md [hunk 1]
```diff
@@ -0,0 +1,184 @@
+# Introduction
+
+## Acknowledgments
+
+This document's section structure is derived from the "[arc42](https://arc42.org/)" architectural template by Dr. Gernot Starke, Dr. Peter Hruschka.
```
Comment (jakobbraun) [inline doc/design.md:5]: I don't get what's acknowleged here. If it's only the structure of a design, I would at least move it to the end of the file.

File: doc/design.md [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+
+### Alternatives considered
+
+1. Checking via HTTP `GET`. Unfortunately this variant is not reliable.
+
+1. Checking all nodes via HTTP `GET`. Suffers from the same problem as the previous idea and additionally requires that the client library knows all data nodes. On top of that, the variant's overhead grows proportionally with the number of nodes.
+
+### Decisions
+
+We decided to base define a monitoring interface that a software that uses the library needs to implement. This allows at least consumers with access to cluster internal information to provide an implementation of this interface.
```
Comment (jakobbraun) [inline doc/design.md:151]:
```suggestion
We decided to define a monitoring interface that a software that uses the library needs to implement. This allows at least consumers with access to cluster internal information to provide an implementation of this interface.
```

File: doc/design.md [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+`dsn~uploading-input-stream-to-bucket~1`
+
+The `Bucket` offers uploading of the contents of an `InputStream` into a file in that bucket on BucketFS.
+
+Covers:
+
+* `req~uploading-input-stream-to-a-file-in-bucketfs~1`
+
+Needs: impl, itest
+
```
Comment (jakobbraun) [inline doc/design.md:97]: I think these design declarations do not add any additional information. I would suggest to leave away design here and directly add `Needs: impl, itest` to the requirements.
In my opinion, this is currently a duplication of the requirements.

**Review by jakobbraun — CHANGES_REQUESTED** [review]

File: doc/design.md (hunk 2 above)
Comment (redcatbear) [reply doc/design.md]: Fixed.

**Review by jakobbraun — APPROVED** [review]

**Pull request merged**